This module gets the dataframe with id, sentiment, canton for instagram json files.

In [1]:
import pandas as pd
import requests
import time
import pickle
import numpy as np
import folium
import scipy.stats as stats
import matplotlib.pyplot as plt
import math
import json
%matplotlib inline

In [2]:
import os
import sys

spark_path = os.environ["SPARK_PATH"]
os.environ['SPARK_HOME'] = spark_path
os.environ['HADOOP_HOME'] = spark_path

sys.path.append(spark_path + "/bin")
sys.path.append(spark_path + "/python")
sys.path.append(spark_path + "/python/pyspark/")
sys.path.append(spark_path + "/python/lib")
sys.path.append(spark_path + "/python/lib/pyspark.zip")
sys.path.append(spark_path + "/python/lib/py4j-0.9-src.zip")

## Defining Spark Context

In [3]:
import json
import subprocess
import fnmatch
from pyspark import SparkConf, SparkContext
from pyspark.sql import SQLContext

# Defining sc
conf = SparkConf().setAppName("ADA-GCL")
sc = SparkContext(conf=conf)

# use SQL context
sqlContext = SQLContext(sc)

## Spark Application

Quite simple at the moment, just getting the dataframe.

In [4]:
# Path definitions
# base_path = 'hdfs://iccluster046.iccluster.epfl.ch:8020/datasets/goodcitylife'
# insta_files='*/harvest3r_instagram_data*.json'
sample_path ='hdfs://iccluster046.iccluster.epfl.ch:8020/datasets/goodcitylife/april/harvest3r_instagram_data_28-04_0.json'

# getting the dataframe
df = sqlContext.read.json(sample_path)

# function
# def get_tags(source):
#     return source['tags']

# df_tags = df.select("_source").foreach(get_tags)

# Displays the content of the DataFrame to stdout
df.show()

+-------------------+--------------------+-----------+--------------------+-------+
|                _id|              _index|     _score|             _source|  _type|
+-------------------+--------------------+-----------+--------------------+-------+
|1461840004439800092|merged_content_20...| 0.18902478|[UNKNOWN,pdeleona...|content|
|1461845617018500058|merged_content_20...| 0.17140923|[UNKNOWN,lolig_sw...|content|
|1461843807915600114|merged_content_20...| 0.16049755|[UNKNOWN,bestworl...|content|
|1461865216725800014|merged_content_20...| 0.13359328|[MALE,netzwandern...|content|
|1461870662929600135|merged_content_20...|  0.1325924|[UNKNOWN,pdeleona...|content|
|1461882421947400206|merged_content_20...|0.122113265|[UNKNOWN,d_e_n_i_...|content|
|1461882806802000025|merged_content_20...| 0.12073777|[UNKNOWN,d_e_n_i_...|content|
|1461883237548800035|merged_content_20...| 0.12073777|[UNKNOWN,d_e_n_i_...|content|
|1461830185754200033|merged_content_20...| 0.08652698|[UNKNOWN,gruyeres...|c

# Local Operations
* will need to move this into Spark Applications with Spark DataFrames, but developing locally using pandas DataFrames now...

### Create a good canton list

In [5]:
geo_map = {'city':['zurich','geneva','lausanne','zermatt','bern','basel','geneve','winterthur','luzern','lucerne','st-gallen','lugano'], 
           'canton': ['Zurich', 'Geneva','Vaud', 'Valais', "Bern", 'Basel-City', 'Geneva', 'Zurich', 'Lucerne','Lucerne', 'St-Gallen', 'Ticino']}

In [6]:
df_geo = pd.DataFrame.from_dict(geo_map)

In [7]:
df_geo.head()

canton      city
0  Zurich    zurich
1  Geneva    geneva
2    Vaud  lausanne
3  Valais   zermatt
4    Bern      bern

In [8]:
top_population = ['zurich','geneva','lausanne','zermatt','bern','basel','geneve','winterthur','luzern','lucerne','st-gallen','lugano','biel','thun','fribourg']
top_tourist = ['zermatt','montreux','jungfrau','interlaken',]

In [9]:
popular_cities = ['zurich','geneva','lausanne','zermatt','bern','basel','geneve','winterthur','luzern','lusern','st-gallen','lugano']

In [10]:
df_geo.city.values

array(['zurich', 'geneva', 'lausanne', 'zermatt', 'bern', 'basel',
       'geneve', 'winterthur', 'luzern', 'lucerne', 'st-gallen', 'lugano'], dtype=object)

### Full Cantons
Downloaded an Excel File

In [11]:
df_cantons = pd.read_csv('cities_cantons.csv')
df_cantons = df_cantons[['Ortschaftsname','Gemeindename','Kantonskürzel']].drop_duplicates()
df_cantons.columns = ['City','Municipality','Canton']
df_cantons.head()

City Municipality Canton
0      Aadorf       Aadorf     TG
1       Aarau        Aarau     AG
3  Aarau Rohr        Aarau     AG
4     Aarberg      Aarberg     BE
5     Aarburg      Aarburg     AG

## Dataframe with Geo and Sentiment
* 1) turning the raw dataframe with id, index, score, source and type into a dataframe with id, tags, and sentiment
* 2) turn the list of geo_id into one new column with just one geolocation
* 3) map that geolocation to another column called canton

### 0) Original DataFrame

In [12]:
df = pd.read_json('Sample Data/harvest3r_instagram_data_15-04_0.json')

In [13]:
df.head()

_id                                   _index    _score  \
0  1460752286000003072  merged_content_2016_04_15_to_2016_04_21  0.048309   
1  1460727556000007168  merged_content_2016_04_15_to_2016_04_21  0.029747   
2  1460728673000006656  merged_content_2016_04_15_to_2016_04_21  0.026873   
3  1460749002000007168  merged_content_2016_04_15_to_2016_04_21  0.022194   
4  1460742655000008960  merged_content_2016_04_15_to_2016_04_21  0.021687   

                                             _source    _type  
0  {'sequence_range': 83081, 'source_link': 'http...  content  
1  {'sequence_range': 70901, 'source_link': 'http...  content  
2  {'sequence_range': 42239, 'source_link': 'http...  content  
3  {'sequence_range': 87456, 'source_link': 'http...  content  
4  {'sequence_range': 58253, 'source_link': 'http...  content

### 1) Getting id, tags, sentiment

In [14]:
def get_tags(source):
    return source['tags']

In [15]:
def get_sentiment(source):
    return source.get('sentiment') # getting the value from key using dictionary

In [16]:
df_sentiment = df._source.apply(get_sentiment).to_frame('Sentiment')

In [17]:
df_tags = df._source.apply(get_tags).to_frame('Tags')

In [18]:
df_extracted = df._id.to_frame('_id').join(df_sentiment)
df_extracted = df_extracted.join(df_tags)
df_extracted

_id Sentiment  \
0     1460752286000003072      None   
1     1460727556000007168      None   
2     1460728673000006656   NEUTRAL   
3     1460749002000007168  POSITIVE   
4     1460742655000008960  POSITIVE   
5     1460742313000008192      None   
6     1460762902000002560   NEUTRAL   
7     1460724976000008960      None   
8     1460859717000004864      None   
9     1460748532000009216      None   
10    1460740192000012544  POSITIVE   
11    1460708866000010240      None   
12    1460700850000008192      None   
13    1460714111000013056   NEUTRAL   
14    1460731443000001536   NEUTRAL   
15    1460686428000010496      None   
16    1460746382000002816  POSITIVE   
17    1460747247000002304   NEUTRAL   
18    1460753510000014080      None   
19    1460711221000014336      None   
20    1460761015000000512      None   
21    1460763857000000256   NEUTRAL   
22    1460750202000013824  POSITIVE   
23    1460713027000000256      None   
24    1460702622000008960      None   
25    1460752487000004352      None   
26    1460702231000014848      None   
27    1460756871000007680      None   
28    1460758695000010752      None   
29    1460738417000014592      None   
...                   ...       ...   
1195  1460758307000002560  POSITIVE   
1196  1460717225000000256      None   
1197  1460708830000010240      None   
1198  1460686153000016128      None   
1199  1460764781000012544  POSITIVE   
1200  1460682872000011008  POSITIVE   
1201  1460718071000003072      None   
1202  1460743032000016128      None   
1203  1460698697000001536      None   
1204  1460740060000008704      None   
1205  1460709645000013056      None   
1206  1460737975000000256      None   
1207  1460711537000012032      None   
1208  1460739277000012032      None   
1209  1460698451000014592      None   
1210  1460735955000010496      None   
1211  1460698451000016128      None   
1212  1460720126000001536  POSITIVE   
1213  1460753390000014080      None   
1214  1460690542000002816      None   
1215  1460750861000003072      None   
1216  1460752684000004352  POSITIVE   
1217  1460718861000012544      None   
1218  1460704331000002048      None   
1219  1460724460000010496  POSITIVE   
1220  1460768850000002816  NEGATIVE   
1221  1460727671000011776      None   
1222  1460745405000001024      None   
1223  1460746515000006656      None   
1224  1460743600000002816      None   

                                                   Tags  
0     [Switzerland, ch, genebra, geneve, genf, ginev...  
1     [art, border, ch, eidgenoss, helvetia, land, m...  
2     [Wintersport, earnyourturns, flylowgear, jungf...  
3     [bahnhofklatscher, eidgenoss, eidgenosse, eidg...  
4     [Aarau, Austria, Austrija, Basel, Bazel, Bern,...  
5     [europa, europe, flag, makeitrain, schweiz, su...  
6     [ahmadiyya, antwort, basel, beatiful, bern, be...  
7     [AtTheAirport2016, blackandwhite, bnw, fujifee...  
8     [CH, bern, citylife, citytrain, commute, europ...  
9     [AtTheAirport2016, blackandwhite, bnw, fujifee...  
10    [all, alles, country, danke, europe, eurowing,...  
11    [awesome, basel, baselschoolofdesign, baselsta...  
12    [arome, basel, bern, bienetre, bonheur, bugies...  
13    [aeroport, choices, geneva, genevaguide, genev...  
14    [Switzerland, Zurich, bahnhofstrasse, ceylon, ...  
15    [holidays, instachile, instaswitzerland, invie...  
16    [allblackeverything, art, beauty, beautyblogge...  
17    [Suiza, Switzerland, arch, archilovers, archit...  
18    [bjj, crossfit, eu, euro, europa, event, fitne...  
19    [april, aprilweather, aprilwetter, cloudy, ig_...  
20    [Nikon, Schweiz, beanie, benetton, crazyabouth...  
21    [AlpineClimbing, Alps, Aperture, COP21, NeverS...  
22    [MasterChef, MasterChefit, bellinzona, foodblo...  
23    [basel, hotel, interlaken, luzern, regen, rest...  
24    [brown, chocolatelab, dawg, dog, dogs, fab_lab...  
25    [Zurich, basel, bern, canvas, download, flower...  
26    [beautiful, bern, clouds, clove

In [19]:
# tags are a list
df_extracted['Tags'][0]

['Switzerland',
 'ch',
 'genebra',
 'geneve',
 'genf',
 'ginevra',
 'onu',
 'schweiz',
 'suisse',
 'svizzera',
 'swiss',
 'un']

### 2) Getting one single city from list of tags

In [20]:
CITY_LIST = df_geo.city.values # gets more complicated as df_geo gets more complicated

def extract_city(tag_list):
    # geo_map['city]
    return next((city for city in tag_list if city in CITY_LIST), None)

In [21]:
df_city = df_extracted.Tags.apply(extract_city).to_frame('City')

In [22]:
df_extracted = df_extracted.join(df_city)

In [23]:
geo_map

{'canton': ['Zurich',
  'Geneva',
  'Vaud',
  'Valais',
  'Bern',
  'Basel-City',
  'Geneva',
  'Zurich',
  'Lucerne',
  'Lucerne',
  'St-Gallen',
  'Ticino'],
 'city': ['zurich',
  'geneva',
  'lausanne',
  'zermatt',
  'bern',
  'basel',
  'geneve',
  'winterthur',
  'luzern',
  'lucerne',
  'st-gallen',
  'lugano']}

In [24]:
geo_map = {'zurich':'Zurich','geneva':'Geneva','lausanne':'Vaud','zermatt':'Valais','bern':'Bern',
           'basel':'Basel-City','geneve':'Geneva','winterthur':'Zurich','luzern':'Lucerne','lucerne':'Lucerne',
          'st-gallen':'St-Gallen','lugano':'Ticino'}

### 3) Mapping City to Canton

In [25]:
df_extracted['Canton'] = df_extracted['City'].map(geo_map)

In [26]:
df_extracted.dropna()

_id Sentiment  \
6    1460762902000002560   NEUTRAL   
10   1460740192000012544  POSITIVE   
13   1460714111000013056   NEUTRAL   
16   1460746382000002816  POSITIVE   
39   1460715820000002304  POSITIVE   
41   1460731351000002048   NEUTRAL   
43   1460709766000001792   NEUTRAL   
50   1460751107000014336  POSITIVE   
52   1460697581000008704   NEUTRAL   
65   1460706032000000256   NEUTRAL   
66   1460720573000010240  POSITIVE   
67   1460720573000010240  POSITIVE   
68   1460720574000013056  POSITIVE   
72   1460703196000002816  POSITIVE   
76   1460756562000013568  POSITIVE   
79   1460757587000016384   NEUTRAL   
90   1460736047000004352  POSITIVE   
91   1460703671000011776  POSITIVE   
128  1460739431000009472  POSITIVE   
129  1460848311000009472  POSITIVE   
141  1460757080000015360  POSITIVE   
143  1460762453000006144   NEUTRAL   
160  1460754691000008960  POSITIVE   
162  1460752156000005120  POSITIVE   
168  1460698770000007168  POSITIVE   
178  1460750641000012544  POSITIVE   
180  1460743288000015104  POSITIVE   
185  1460761775000004864  POSITIVE   
186  1460743288000015104  POSITIVE   
202  1460701785000010240  POSITIVE   
..                   ...       ...   
832  1460732723000007168   NEUTRAL   
837  1460723391000000256  POSITIVE   
841  1460697971000007424   NEUTRAL   
850  1460697972000007424   NEUTRAL   
857  1460697971000007424   NEUTRAL   
858  1460697971000007424   NEUTRAL   
859  1460697972000007424   NEUTRAL   
860  1460687874000005888  POSITIVE   
865  1460697971000007424   NEUTRAL   
867  1460721375000010240  POSITIVE   
873  1460827972000004352  POSITIVE   
877  1460728837000013056  POSITIVE   
880  1460697972000008192   NEUTRAL   
883  1460700756000001536   NEUTRAL   
887  1460753100000003840   NEUTRAL   
891  1460728836000010240   NEUTRAL   
893  1460745905000007424  POSITIVE   
895  1460707011000016128  POSITIVE   
901  1460886045000014592  POSITIVE   
906  1460738350000015616  POSITIVE   
912  1460737096000013056  POSITIVE   
921  1460718685000002816  POSITIVE   
922  1460707476000002304  POSITIVE   
936  1460748756000003584  POSITIVE   
937  1460707885000013056  POSITIVE   
941  1460751897000006400  POSITIVE   
950  1460737052000011008  POSITIVE   
951  1460707885000014336  POSITIVE   
952  1460734581000012032  POSITIVE   
959  1460707885000014336  POSITIVE   

                                                  Tags      City      Canton  
6    [ahmadiyya, antwort, basel, beatiful, bern, be...     basel  Basel-City  
10   [all, alles, country, danke, europe, eurowing,...    zurich      Zurich  
13   [aeroport, choices, geneva, genevaguide, genev...    geneva      Geneva  
16   [allblackeverything, art, beauty, beautyblogge...    geneve      Geneva  
39   [aeroport, airport, beautiful, bigworld, board...    geneva      Geneva  
41   [djpush, matterhorn, pjsweetcouple, push_dj, p...   zermatt      Valais  
43   [badboy, badboys, bijoux, bikerlife, booba, ch...    geneve      Geneva  
50   [black, blackandwhite, blackandwhitephotograph...    geneva      Geneva  
52   [Butterfly, KATRINLANGER, Lucerne, accessories...    luzern     Lucerne  
65   [beautiful, bling, celiabrown, celiabrownjewel...    geneva      Geneva  
66   [and, ataxia, booyah, club, dancefloor, dims, ...    geneva      Geneva  
67   [and, ataxia, booyah, club, dancefloor, dims, ...    geneva      Geneva  
68   [and, ataxia, booyah, club, dancefloor, dims, ...    geneva      Geneva  
72   [australia, banjaluka, beograd, bijeljina, bre...    lugano      Ticino  
76   [brunette, but, curves, curvy, diffrent, frank...    luzern     Lucerne  
79   [YGREK, YGREK1, art, blackandwhite, blackandwh...    geneva      Geneva  
90   [accessoires, befit, body, brieftasche, busine...    zurich      Zurich  
91   [accessoires, befit, body, brieftasche, busine...    zurich      Zurich  
128  [beauty, beautyblogger, browonfleek, celebrity...    geneva      Geneva  
129  [carouge, geneva, geneva2016, genevafood, gene...    geneva      Geneva  
141  

In [27]:
df_extracted.shape[0]

1225

Only 123 out of 1225 have both sentiment and canton. It's just 10% now. We need to get more sentiment using CNN and get more geolocation by using a more comprehensive dictionary of cities and cantons. The current manual list fills about 33% of total locations.